lib/SIL/OwnershipUtils.cpp

```diff
@@ -438,3 +387,128 @@ bool BorrowScopeIntroducingValue::visitLocalScopeTransitiveEndingUses(
 
   return foundError;
 }
+
+//===----------------------------------------------------------------------===//
+// Searching for BorrowScopeIntroducers in Def-Use Graph
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// A generator of borrow scope introducing values.
+struct BorrowScopeIntroducingValueGenerator {
+  SmallVector<SILValue, 32> worklist;
+
+  BorrowScopeIntroducingValueGenerator(SILValue value) {
+    assert(value.getOwnershipKind() == ValueOwnershipKind::Guaranteed);
+    worklist.push_back(value);
+  }
+
+  class Result {
+    Optional<BorrowScopeIntroducingValue> value;
+    bool hadError;
+
+  public:
+    Result(BorrowScopeIntroducingValue value) : value(value), hadError(false) {}
+    Result(llvm::NoneType) : value(), hadError(true) {}
+    Result() : value(), hadError(false) {}
+
+    bool finishedDueToError() const { return !value.hasValue() && hadError; }
+    bool finishedDueToNormalCompletion() const {
+      return !value.hasValue() && !hadError;
+    }
+
+    bool hasValue() const { return value.hasValue(); }
+    BorrowScopeIntroducingValue getValue() const { return value.getValue(); }
+  };
+
+  Result next();
+};
+
+} // end anonymous namespace
+
+BorrowScopeIntroducingValueGenerator::Result
+BorrowScopeIntroducingValueGenerator::next() {
+  while (!worklist.empty()) {
+    SILValue v = worklist.pop_back_val();
+
+    // First check if v is an introducer. If so, stash it and continue.
+    if (auto scopeIntroducer = BorrowScopeIntroducingValue::get(v)) {
+      return *scopeIntroducer;
+    }
+
+    // If v produces .none ownership, then we can ignore it. It is important
+    // that we put this before checking for guaranteed forwarding instructions,
+    // since we want to ignore guaranteed forwarding instructions that in this
+    // specific case produce a .none value.
+    if (v.getOwnershipKind() == ValueOwnershipKind::None)
+      continue;
+
+    // Otherwise if v is an ownership forwarding value, add its defining
+    // instruction
+    if (isGuaranteedForwardingValue(v)) {
+      if (auto *i = v->getDefiningInstruction()) {
+        llvm::transform(i->getAllOperands(), std::back_inserter(worklist),
```

Comment:
Any forwarding instruction should identify the single operand that it forwards. We shouldn't just arbitrarily dump all the operands on a worklist.

Reply:
It is nonsense to forward multiple guaranteed values into a single scope.